### ISSUE TITLE
Support bitfields in low-level bindings

### ISSUE THREAD

**Issue by edsko** [well-typed/hs-bindgen#17]

**Comment by TravisCardwell** [comment]
[C99][] section 6.7.2.1 paragraph 10:

[C99]: <https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf>

> An implementation may allocate any addressable storage unit large enough to hold a bit-field. If enough space remains, a bit-field that immediately follows another bit-field in a structure shall be packed into adjacent bits of the same unit. If insufficient space remains, whether a bit-field that does not fit is put into the next unit or overlaps adjacent units is implementation-defined. The order of allocation of bit-fields within a unit (high-order to low-order or low-order to high-order) is implementation-defined. The alignment of the addressable storage unit is unspecified.

When we generate `Storable` instances, field offsets are for the specific version of Clang/LLVM used.  If users build with a different compiler/version, perhaps our generated `Storable` instances could be inconsistent?  Should we take measures to address this?  Our generated tests should catch it, at least.

**Comment by TravisCardwell** [comment]
[C99](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf) section 6.7.2 paragraph 5:

> Each of the comma-separated sets designates the same type, except that for bit-fields, it is implementation-defined whether the specifier `int` designates the same type as `signed int` or the same type as `unsigned int`.

I was thinking about how we might implement domain checks when setting bit field values, and this complicates things!

Can we determine the minimum and maximum values for an `int` bit field?  I do not see a way to do so using `libclang`.  Perhaps the signedness for the implementation can be determined using a `configure`-style test.

<details><summary>Experiment</summary>

```c
#include <stdio.h>

typedef struct foo {
  int a : 3;
} foo;

int main() {
  foo x = { .a = -1 };
  printf("%d\n", (long) x.a);
  printf("signed: %s\n", (long) x.a < 0 ? "yes" : "no");
  return 0;
}
```

</details>

**Comment by phadej** [comment]
AFAICT, gcc just takes the lower bits which fit the field, and doesn't care about signedness whatsover

```c
#include <stdio.h>

union record {
        struct {
                unsigned int lo : 8;
                signed int hi : 8;
        };
        int full;
};

int main() {
        union record r;
        int demo;

        r.lo = 0xffed;
        r.hi = 0xffcb;
        printf("%x %x %x\n", r.lo, r.hi, r.full);

        demo = -1000;
        r.hi = demo;
        printf("%x %x %x %x\n", r.lo, demo, r.hi, r.full);

        demo = -10;
        r.hi = demo;
        printf("%x %x %x %x\n", r.lo, demo, r.hi, r.full);

        demo = 0xfff1ff80;
        r.hi = demo;
        printf("%x %x %x %x\n", r.lo, demo, r.hi, r.full);
}
```

which prints

```c
ed ffffffcb d200cbed
ed fffffc18 18 d20018ed
ed fffffff6 fffffff6 d200f6ed
ed fff1ff80 ffffff80 d20080ed
```

so the only "weird" thing happening, is that it seems that for the signed fields the [sign is extended](https://en.wikipedia.org/wiki/Sign_extension) on read, which makes values which fit come back negative too.

I'll guess for bitfields the type has always to be primitive, and we know whether the field is signed or unsigned; so we can do sign extension for signed ones.

Interesting corner case, but AFAICT nothing too complicated is happening at least on x86_64.

**Comment by phadej** [comment]
>  If users build with a different compiler/version, perhaps our generated Storable instances could be inconsistent? 

That's true for all aspects. We generate implementation specific code. Bitfields just happen to be part where different compilers indeed may do different things even on the same platform. Standard doesn't dictate ABI on any platform, it's all implementation specific.

**Comment by phadej** [comment]
But that said, part of that ABI-specifity is on us. By default GHC uses GCC as the C-compiler e.g. on linux. I have no idea whether it can be made to be `clang` *and what actually breaks then*, as `gcc` is quite hardcoded (e.g. i'm 99% sure that if `Cabal` still invokes C-compiler manually, it doesn't respect `ghc --info` fields, and just uses `gcc`).

**Comment by phadej** [comment]
This is done in #410
